### PR TITLE
Query: Adds hybrid search query pipeline stage

### PR DESCRIPTION
**Pull request by neildsh**
## Description

Adds hybrid search query pipeline stage. This requires the new Direct package and gateway to be available in order to light up.

Given an input SQL such as:
```sql
      SELECT TOP 100 c.text, c.abstract
      FROM c
      ORDER BY RANK RRF(FullTextScore(c.text, ['swim', 'run']), FullTextScore(c.abstract, ['energy']))
```

The new query plan (encoded below as XML instead of JSON to help readability) is as follows:

```
        <queryRanges>
          <Item>{"min":[],"max":"Infinity","isMinInclusive":true,"isMaxInclusive":false}</Item>
        </queryRanges>
        <hybridSearchQueryInfo>
          <globalStatisticsQuery><![CDATA[
SELECT 
    COUNT(1) AS documentCount,
    [
        {
            totalWordCount: SUM(_FullTextWordCount(c.text)),
            hitCounts: [
                COUNTIF(FullTextContains(c.text, "swim")),
                COUNTIF(FullTextContains(c.text, "run"))
            ]
        },
        {
            totalWordCount: SUM(_FullTextWordCount(c.abstract)),
            hitCounts: [
                COUNTIF(FullTextContains(c.abstract, "energy"))
            ]
        }
    ] AS fullTextStatistics
FROM c
]]></globalStatisticsQuery>
          <componentQueryInfos>
            <Item>
              <distinctType>None</distinctType>
              <top>200</top>
              <orderBy>
                <Item>Descending</Item>
              </orderBy>
              <orderByExpressions>
                <Item>_FullTextScore(c.text, ["swim", "run"], {documentdb-formattablehybridsearchquery-totaldocumentcount}, {documentdb-formattablehybridsearchquery-totalwordcount-0}, {documentdb-formattablehybridsearchquery-hitcountsarray-0})</Item>
              </orderByExpressions>
              <hasSelectValue>false</hasSelectValue>
              <rewrittenQuery><![CDATA[
SELECT TOP 200 
    c._rid,
    [
        {
            item: _FullTextScore(c.text, ["swim", "run"], {documentdb-formattablehybridsearchquery-totaldocumentcount}, {documentdb-formattablehybridsearchquery-totalwordcount-0}, {documentdb-formattablehybridsearchquery-hitcountsarray-0})
        }
    ] AS orderByItems,
    {
        payload: {
            text: c.text,
            abstract: c.abstract
        },
        componentScores: [
            _FullTextScore(c.text, ["swim", "run"], {documentdb-formattablehybridsearchquery-totaldocumentcount}, {documentdb-formattablehybridsearchquery-totalwordcount-0}, {documentdb-formattablehybridsearchquery-hitcountsarray-0}),
            _FullTextScore(c.abstract, ["energy"], {documentdb-formattablehybridsearchquery-totaldocumentcount}, {documentdb-formattablehybridsearchquery-totalwordcount-1}, {documentdb-formattablehybridsearchquery-hitcountsarray-1})
        ]
    } AS payload
FROM c
WHERE {documentdb-formattableorderbyquery-filter}
ORDER BY _FullTextScore(c.text, ["swim", "run"], {documentdb-formattablehybridsearchquery-totaldocumentcount}, {documentdb-formattablehybridsearchquery-totalwordcount-0}, {documentdb-formattablehybridsearchquery-hitcountsarray-0}) DESC
]]></rewrittenQuery>
              <hasNonStreamingOrderBy>true</hasNonStreamingOrderBy>
            </Item>
            <Item>
              <distinctType>None</distinctType>
              <top>200</top>
              <orderBy>
                <Item>Descending</Item>
              </orderBy>
              <orderByExpressions>
                <Item>_FullTextScore(c.abstract, ["energy"], {documentdb-formattablehybridsearchquery-totaldocumentcount}, {documentdb-formattablehybridsearchquery-totalwordcount-1}, {documentdb-formattablehybridsearchquery-hitcountsarray-1})</Item>
              </orderByExpressions>
              <hasSelectValue>false</hasSelectValue>
              <rewrittenQuery><![CDATA[
SELECT TOP 200 
    c._rid,
    [
        {
            item: _FullTextScore(c.abstract, ["energy"], {documentdb-formattablehybridsearchquery-totaldocumentcount}, {documentdb-formattablehybridsearchquery-totalwordcount-1}, {documentdb-formattablehybridsearchquery-hitcountsarray-1})
        }
    ] AS orderByItems,
    {
        payload: {
            text: c.text,
            abstract: c.abstract
        },
        componentScores: [
            _FullTextScore(c.text, ["swim", "run"], {documentdb-formattablehybridsearchquery-totaldocumentcount}, {documentdb-formattablehybridsearchquery-totalwordcount-0}, {documentdb-formattablehybridsearchquery-hitcountsarray-0}),
            _FullTextScore(c.abstract, ["energy"], {documentdb-formattablehybridsearchquery-totaldocumentcount}, {documentdb-formattablehybridsearchquery-totalwordcount-1}, {documentdb-formattablehybridsearchquery-hitcountsarray-1})
        ]
    } AS payload
FROM c
WHERE {documentdb-formattableorderbyquery-filter}
ORDER BY _FullTextScore(c.abstract, ["energy"], {documentdb-formattablehybridsearchquery-totaldocumentcount}, {documentdb-formattablehybridsearchquery-totalwordcount-1}, {documentdb-formattablehybridsearchquery-hitcountsarray-1}) DESC
]]></rewrittenQuery>
              <hasNonStreamingOrderBy>true</hasNonStreamingOrderBy>
            </Item>
          </componentQueryInfos>
          <take>100</take>
          <requiresGlobalStatistics>true</requiresGlobalStatistics>
        </hybridSearchQueryInfo>
```

We have a custom implementation for the global statistics inside the `HybridSearchCrossPartitionQueryPipelineStage` because it uses nested aggregates. Each of the component queries in the hybrid search query plan is cross partition, and we run them using the existing cross partition query pipelines.

Note the use of placeholders such as `{documentdb-formattablehybridsearchquery-totaldocumentcount}` in the query plan. These need to be replaced by the global statistics.

## Type of change

- [x] New feature (non-breaking change which adds functionality)

### CODE REVIEW

**Comment by sboshra** [comment]
>             {

disable ODE for hybrid search queries 

---
Refers to: Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CosmosQueryExecutionContextFactory.cs:282 in fa9e5ff. [](commit_id = fa9e5ff9d80881ead707560c7a20b1b59c9d8071, deletion_comment = False)

File: Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/HybridSearch/FullTextStatistics.cs [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    using System.Collections.Generic;
+    using Microsoft.Azure.Cosmos.CosmosElements;
+
+    internal sealed class FullTextStatistics
+    {
+        public long TotalWordCount { get; }
+
+        public ReadOnlyMemory<long> HitCounts => this.hitCounts;
+
+        private readonly long[] hitCounts;
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/HybridSearch/FullTextStatistics.cs:17]: style?

File: Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/HybridSearch/HybridSearchCrossPartitionQueryPipelineStage.cs [hunk 2]
```diff
@@ -0,0 +1,1052 @@
...
+    using Microsoft.Azure.Cosmos.Query.Core.QueryClient;
+    using Microsoft.Azure.Cosmos.Query.Core.QueryPlan;
+    using Microsoft.Azure.Cosmos.Tracing;
+    using Microsoft.Azure.Documents;
+
+    internal class HybridSearchCrossPartitionQueryPipelineStage : IQueryPipelineStage
+    {
+        private delegate TryCatch<IQueryPipelineStage> HybridSearchComponentPipelineFactory(QueryInfo queryInfo);
+
+        private const string DisallowContinuationTokenMessage = "Hybrid search does not support continuation tokens";
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/HybridSearch/HybridSearchCrossPartitionQueryPipelineStage.cs:30]: >tokens [](http://example.com/codeflow?start=109&length=6)

missing '.'

File: Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/HybridSearch/HybridSearchCrossPartitionQueryPipelineStage.cs [hunk 3]
```diff
@@ -0,0 +1,1052 @@
...
+
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/HybridSearch/HybridSearchCrossPartitionQueryPipelineStage.cs:31]: Continuation tokens are not...

Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/HybridSearch/HybridSearchCrossPartitionQueryPipelineStage.cs:31]: Please check if all error messages could be pulled into a resource file

File: Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/HybridSearch/HybridSearchCrossPartitionQueryPipelineStage.cs [hunk 4]
```diff
@@ -0,0 +1,1052 @@
...
+        private const int RRFConstant = 60;
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/HybridSearch/HybridSearchCrossPartitionQueryPipelineStage.cs:32]: >RRFConstant [](http://example.com/codeflow?start=26&length=11)

RffConstant

File: Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/HybridSearch/HybridSearchCrossPartitionQueryPipelineStage.cs [hunk 5]
```diff
@@ -0,0 +1,988 @@
...
+            if (queryResults.Count == 0 || queryResults.Count == 1)
+            {
+                return TryCatch<(IReadOnlyList<HybridSearchQueryResult>, QueryPage emptyPage)>.FromResult((queryResults, emptyPage));
+            }
+
+            HybridSearchDebugTraceHelpers.TraceQueryResults(queryResults, queryPipelineStages.Count);
+
+            queryResults.Sort((x, y) => string.CompareOrdinal(x.Rid.Value, y.Rid.Value));
+
+            UniqueRids(queryResults);
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/HybridSearch/HybridSearchCrossPartitionQueryPipelineStage.cs:436]: >UniqueRids [](http://example.com/codeflow?start=12&length=10)

Consider a verb for the name

File: Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/HybridSearch/HybridSearchCrossPartitionQueryPipelineStage.cs [hunk 6]
```diff
@@ -0,0 +1,988 @@
...
+            IReadOnlyList<List<ScoreTuple>> componentScores = tryGetComponentScores.Result;
+
+            foreach (List<ScoreTuple> scoreTuples in componentScores)
+            {
+                scoreTuples.Sort((x, y) => (-1) * x.Score.CompareTo(y.Score)); // sort descending, since higher scores are better
+            }
+
+            int[,] ranks = ComputeRanks(componentScores);
+
+            ComputeRRFScores(ranks, queryResults);
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/HybridSearch/HybridSearchCrossPartitionQueryPipelineStage.cs:453]: >ComputeRRFScores [](http://example.com/codeflow?start=12&length=16)

ComputeRrfScores

**Review by sboshra — DISMISSED** [review]
:shipit:

File: Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/HybridSearch/HybridSearchCrossPartitionQueryPipelineStage.cs [hunk 7]
```diff
@@ -0,0 +1,988 @@
...
+            foreach (IQueryPipelineStage queryPipelineStage in queryPipelineStages)
+            {
+                prefetchers.Add(new QueryPipelineStagePrefetcher(queryPipelineStage));
+            }
+
+            await ParallelPrefetch.PrefetchInParallelAsync(prefetchers, maxConcurrency, trace, cancellationToken);
+
+            double requestCharge = 0;
+            QueryPageParameters queryPageParameters = null;
+            List<HybridSearchQueryResult> queryResults = new List<HybridSearchQueryResult>();
```
Comment (neildsh) [inline Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/HybridSearch/HybridSearchCrossPartitionQueryPipelineStage.cs:480]: >List<HybridSearchQueryResult>(); [](http://example.com/codeflow?start=61&length=32)

Consider initializing this with the expected count of results

**Review by sboshra — APPROVED** [review]

**Review by sc978345 — APPROVED** [review]

**Pull request merged**